pallets/payment-streams/src/utils.rs

```diff
@@ -0,0 +1,490 @@
+    /// As such, the last charged proof can't be greater than the last valid proof, and if they are equal then no charge is made.
+    ///
+    /// Note: right now, the other functions receive the BSP's account ID because it is needed to transfer the user payment to it in this function, but the idea is to change the
+    /// BSP struct to hold a `payment_address` where the payment should be received, and then the functions of this pallet will receive the BSP's ID instead.
+    pub fn do_charge_payment_stream(
+        bsp_account: &T::AccountId,
+        user_account: &T::AccountId,
+    ) -> Result<BalanceOf<T>, DispatchError> {
+        // Get the BSP ID of the BSP account
+        let bsp_id = <T::Providers as storage_hub_traits::ProvidersInterface>::get_provider(
+            bsp_account.clone(),
+        )
+        .ok_or(Error::<T>::NotABackupStorageProvider)?;
+
+        // Check that a payment stream between that BSP and user exists
+        ensure!(
+            PaymentStreams::<T>::contains_key(bsp_id, user_account),
+            Error::<T>::PaymentStreamNotFound
+        );
+
+        // Get the information of the payment stream
+        let payment_stream = PaymentStreams::<T>::get(bsp_id, user_account)
+            .ok_or(Error::<T>::PaymentStreamNotFound)?;
+
+        // Note: No need to check if a new proof has been submitted since the last charge as the only consecuence of that is charging 0 to the user,
+        // and not erroring out helps to be able to call this function without errors when updating or removing a payment stream.
+
+        // Calculate the time passed between the last valid proof and the last charged proof
+        let time_passed = expect_or_err!(payment_stream
+            .last_valid_proof
+            .checked_sub(&payment_stream.last_charged_proof), "Last valid proof should always be greater than or equal to the last charged proof, inconsistency error.",
+            Error::<T>::LastChargeGreaterThanLastValidProof);
+
+        // Convert it to the balance type (for math)
+        let time_passed_balance_typed = T::BlockNumberToBalance::convert(time_passed);
+
+        // Calculate the amount to charge
+        let amount_to_charge = payment_stream
+            .rate
+            .checked_mul(&time_passed_balance_typed)
+            .ok_or(Error::<T>::ChargeOverflow)?;
+
+        // Check the free balance of the user
+        let user_balance = T::NativeBalance::reducible_balance(
+            &user_account,
+            Preservation::Preserve,
+            Fortitude::Polite,
+        );
+
+        // If the user does not have enough balance to pay for its storage:
+        if user_balance < amount_to_charge {
+            // Flag it in the UsersWithoutFunds mapping and emit the UserWithoutFunds event
+            UsersWithoutFunds::<T>::insert(user_account, ());
```

Comment:
In this case, shouldn't we charge whatever he has left? Not only it is more fair, but also the way it is right now, the user might not have enough funds to pay to an SP he owes 1M DOT, but another SP with whom he also has a payment stream, might be able to charge him, let's say the 1k DOT he has left.

Reply:
Yep. I'll work on the out-of-funds logic in another PR. Agree with everything you just said here